mithril-common/src/resource_pool.rs

```diff
@@ -0,0 +1,316 @@
+//! Resource pool implementation
+
+use anyhow::Context;
+use std::{
+    collections::VecDeque,
+    ops::{Deref, DerefMut},
+    sync::{Condvar, Mutex},
+    time::Duration,
+};
+use thiserror::Error;
+
+use crate::StdResult;
+
+/// [ResourcePool] related errors.
+#[derive(Error, Debug)]
+pub enum ResourcePoolError {
+    /// Internal Mutex is poisoned
+    #[error("Poisoned mutex caused error during acquire lock on resource pool")]
+    PoisonedLock(),
+
+    /// Acquire resource has timed out
+    #[error("Acquire resource has timed out")]
+    AcquireTimeout(),
+}
+
+/// Resource pool implementation (FIFO)
+pub struct ResourcePool<T: Send + Sync> {
+    /// The size of the pool
+    size: usize,
+
+    /// Discriminant for the resource pool to check if a returned resource is stale
+    discriminant: Mutex<u64>,
+
+    /// Resources in the pool
+    resources: Mutex<VecDeque<T>>,
+
+    /// Condition variable to notify when a resource is available
+    not_empty: Condvar,
+}
+
+impl<T: Send + Sync> ResourcePool<T> {
+    /// Create a new resource pool
+    pub fn new(pool_size: usize, resources: Vec<T>) -> Self {
+        Self {
+            size: pool_size,
+            discriminant: Mutex::new(0),
+            resources: Mutex::new(resources.into()),
+            not_empty: Condvar::new(),
+        }
+    }
+
+    /// Acquire a resource from the pool with a timeout
+    pub fn acquire_resource(&self, timeout: Duration) -> StdResult<ResourcePoolItem<'_, T>> {
+        let mut resources = self
+            .resources
+            .lock()
+            .map_err(|_| ResourcePoolError::PoisonedLock())
+            .with_context(|| "Resource pool 'acquire_resource' failed locking Mutex")?;
+        while resources.is_empty() {
+            let (resources_locked, timeout) =
+                self.not_empty.wait_timeout(resources, timeout).unwrap();
+            if timeout.timed_out() {
+                return Err(ResourcePoolError::AcquireTimeout())
+                    .with_context(|| "Resource pool 'acquire_resource' has timed out");
+            }
+            resources = resources_locked;
+        }
+
+        Ok(ResourcePoolItem::new(self, resources.pop_front().unwrap()))
+    }
+
+    /// Return a resource to the pool
+    /// A resource is returned to the pool only if the discriminant matches
+    /// and if the pool is not already full
+    pub fn return_resource(&self, resource: T, discriminant: u64) -> StdResult<()> {
```

Comment:
`return` is the (key)word use to return a value from a function.
 Using this word can be misinterpreted and suggest that the function return a value. 
 Can we found a synonym ? `put back`  ?

Reply:
Done